### ISSUE TITLE
BatchNorm initialization

### ISSUE THREAD

**Issue by ZagButNoZig** [patrick-kidger/equinox#955]
First of all, thank you for your great library!

### The problem

- Initialize a model with multiple batch norms (e.g. ResNet)
- Set model to inference mode
- Observe numerical instability (values > 10^9 after a few ResNet-Blocks)

If I read your code correctly variance in BatchNorm is initialized [to zero](https://github.com/patrick-kidger/equinox/blob/main/equinox/nn/_batch_norm.py#L97), meaning that if you use BatchNorm in inference without any training (think: testing random model performance on dataset or the first iteration of an RL algorithm), the BatchNorm scales all values by $\frac{1}{\sqrt{eps}} = 10^4$. Both [PyTorch](https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/nn/modules/batchnorm.py#L66) and [Flax](https://github.com/google/flax/blob/dd6e5955aa6319f9b4fe94bf3c8267bae0e18697/flax/nnx/nn/normalization.py#L288) use $\sigma^2 = 1$ in their code. I think this is a much saner default, considering that $\sigma^2 = 0$ doesn't really make sense.

Tl;dr:

- BatchNorm causes instability due to the initial value of the variance
- Both Flax and PyTorch use variance = 1 as their initial value

**Comment by patrick-kidger** [comment]
I think that's a good spot! Tagging @lockwo and #948 here -- we're thinking of reworking BatchNorm. @lockwo WDYT?

**Comment by lockwo** [comment]
Yea, that's a good point. It has less effect on the `batch` mode of the PR (since the weighting is all based on the current batch), but will still impact evaluation. I changed the initialization to be ones for the std in that case.

I can also change it for the existing approach (ema), but that would result in a change in what batch norm does (and while it likely results in better performance in many cases, would change how default batch norm would behave, which I'm not sure would meet the requirement of "bit for bit backwards compatibility" https://github.com/patrick-kidger/equinox/pull/675#discussion_r1536698496, but we could just put a note saying if you want the old init, you can run this tree_at command).

**Comment by ZagButNoZig** [comment]
> I can also change it for the existing approach (ema), but that would result in a change in what batch norm does (and while it likely results in better performance in many cases, would change how default batch norm would behave, which I'm not sure would meet the requirement of "bit for bit backwards compatibility" https://github.com/patrick-kidger/equinox/pull/675#discussion_r1536698496, but we could just put a note saying if you want the old init, you can run this tree_at command).

I would argue that this is a bugfix that is backwards compatible, but that's just my 2 cents.
